alibi_detect/od/pytorch/pca.py

```diff
@@ -0,0 +1,237 @@
+from typing import Optional, Union, Callable
+
+import torch
+
+from alibi_detect.od.pytorch.base import TorchOutlierDetector
+
+
+class PCATorch(TorchOutlierDetector):
+    ensemble = False
+
+    def __init__(
+            self,
+            n_components: int,
+            device: Optional[Union[str, torch.device]] = None
+            ):
+        """PyTorch backend for PCA detector.
+
+        Parameters
+        ----------
+        n_components:
+            The number of dimensions in the principle subspace. For linear PCA should have
+            ``1 <= n_components < dim(data)``. For kernel pca should have ``1 <= n_components < len(data)``.
+        device
+            Device type used. The default tries to use the GPU and falls back on CPU if needed. Can be specified by
+            passing either ``'cuda'``, ``'gpu'``, ``'cpu'`` or an instance of ``torch.device``.
+        """
+        super().__init__(device=device)
+        self.n_components = n_components
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Detect if `x` is an outlier.
+
+        Parameters
+        ----------
+        x
+            `torch.Tensor` with leading batch dimension.
+
+        Returns
+        -------
+        `torch.Tensor` of ``bool`` values with leading batch dimension.
+
+        Raises
+        ------
+        ThresholdNotInferredException
+            If called before detector has had `infer_threshold` method called.
+        """
+        scores = self.score(x)
+        if not torch.jit.is_scripting():
+            self.check_threshold_inferred()
+        preds = scores > self.threshold
+        return preds.cpu()
+
+    def score(self, x: torch.Tensor) -> torch.Tensor:
+        """Computes the score of `x`
+
+        Parameters
+        ----------
+        x
+            The tensor of instances. First dimension corresponds to batch.
+
+        Returns
+        -------
+        Tensor of scores for each element in `x`.
+
+        Raises
+        ------
+        NotFitException
+            If called before detector has been fit.
+        """
+        self.check_fitted()
+        score = self.compute_score(x)
+        return score.cpu()
+
+    def fit(self, x_ref: torch.Tensor) -> None:
+        """Fits the PCA detector.
+
+        Parameters
+        ----------
+        x_ref
+            The Dataset tensor.
+        """
+        self.x_ref_mean = x_ref.mean(0)
+        self.pcs = self.compute_pcs(x_ref)
+        self.x_ref = x_ref
```

Comment:
Is it necessary to store `x_ref`?

Reply:
(For linear variant)

Reply:
Yeah, good point, I've moved to the kernelPCA class!